### PR TITLE
ci(coverage): add total typescript code coverage statistics

### LINKED ISSUE #2661: ci(coverage): add total typescript code coverage statistics
### Description

As a maintainer I want to know if we have a reasonable amount of coverage so that I'm more confident in the production readiness of the codebase.

Note: 'reasonable amount' is highly subjective. This issue is not about asserting what that percentage is because it varies. My personal opinion is that critical code-paths need 100% code coverage, especially for catch blocks (error handling) while non-critical code-paths can get away with the bare minimum or no coverage at all.

More context:
We have used https://about.codecov.io/ in the past but then we had to break the test execution up into several (dozens) of separate GH Action Workflow jobs and we can no longer measure total code coverage as of right now.
This issue is about figuring out how to remedy that and get us back on track with a metric that we can use on a badge in the readme.

**Additional context on 2024-02-15**

The steps I'd take to solving this:
1. Research and figure out how can coverage reports be either merged or separately uploaded to codecov and then have codecov merge the statistics themselves. The latter is preferred if achievable, but the former is the likely outcome.
2. Once the merging is figured out, we can update the ci tasks that are running Jest tests to do the uploading of the statistics  to codecov.
3. Screenshots below showing what are the files that are being generated when `--coverage` is passed in to Jest

image

image

### Acceptance Criteria
1. Total code coverage metric re-introduced.
4. No major refactor on the CI needed. This is important because there are already other factors at play about how we structure the jobs/workflows that are more important (performance which is $$$ because of  billed hours the CI has to run)

### PR DESCRIPTION
**Pull request by aldousalvarez**
### **Commit** to be reviewed
---
ci(coverage): add total typescript code coverage statistics
```
Primary Changes
----------------
1. Updated the ci.yaml and ci.sh to introduce total typescript code coverage statistics using jest and istanbul-merge
```
Fixes #2661

**Pull Request Requirements**
- [x] Rebased onto `upstream/main` branch and squashed into single commit to help maintainers review it more efficient and to avoid spaghetti git commit graphs that obfuscate which commit did exactly what change, when and, why.
- [x] Have git sign off at the end of commit message to avoid being marked red. You can add `-s` flag when using `git commit` command. You may refer to this [link](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) for more information.
- [x] Follow the Commit Linting specification. You may refer to this [link](https://www.conventionalcommits.org/en/v1.0.0-beta.4/#specification) for more information. 

**Character Limit**
- [x] Pull Request Title and Commit Subject must not exceed 72 characters (including spaces and special characters).
- [x] Commit Message per line must not exceed 80 characters (including spaces and special characters).

**A Must Read for Beginners**
For rebasing and squashing, here's a [must read guide](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing) for beginners.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -354,6 +365,8 @@ jobs:
       FULL_BUILD_DISABLED: true
       JEST_TEST_PATTERN: packages/cactus-api-client/src/test/typescript/(unit|integration|benchmark)/.*/*.test.ts
       JEST_TEST_RUNNER_DISABLED: false
+      JEST_TEST_COVERAGE_PATH: ./code-coverage-ts/cactus-api-client
+      JEST_TEST_CODE_COVERAGE_ENABLED: false
```
Comment (petermetz) [inline .github/workflows/ci.yaml:369]: @aldousalvarez Where is this flag is being implemented for it to be effective?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -2303,6 +2706,56 @@ jobs:
...
+        with:
+          image-ref: 'cactus-quorum-multi-party-all-in-one'
+          format: 'table'
+          exit-code: '1'
+          ignore-unfixed: true
+          vuln-type: 'os,library'
+          severity: 'CRITICAL,HIGH'
+  generate_coverage_report:
+    runs-on: ubuntu-22.04
+    needs: [check-coverage, cp-bungee-hermes]
```
Comment (petermetz) [inline .github/workflows/ci.yaml:2720]: @aldousalvarez Why does this specifically depend on `cp-bungee-hermes` what makes that package special compard to all the other packages we are also testing in this workflow?

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -2303,6 +2706,56 @@ jobs:
...
+        path: ./code-coverage-ts/
+
+    - name: Merge and generate coverage reports
+      run: |
+        ls -R ./code-coverage-ts/**/**/coverage-final.json
+        npx istanbul-merge --out coverage.json ./code-coverage-ts/coverage-reports-*/*/coverage-final.json
+        npx istanbul report --include coverage.json --dir cacti html
+        npx istanbul report --include coverage.json --dir cacti text
+        npx istanbul report --include coverage.json --dir cacti text-summary
+
```
Comment (petermetz) [inline .github/workflows/ci.yaml:2756]: @aldousalvarez I see that this generates the final, combined report, but does it also get uploaded to [CodeCov](https://about.codecov.io/) which would then propgate to the codecov badge on the root README.md file?

I thought we'd need something like this as the final step to make sure that the code coverage file gets uploaded (and also to be able to verify that it's in the valid format accepted by codecov):

https://github.com/codecov/codecov-action

**Review by petermetz — COMMENTED** [review]
@aldousalvarez Thank you very much for this! It's looking very good so far, I just have some questions above^^

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (aldousalvarez) [reply .github/workflows/ci.yaml:369]: @petermetz It is used on the ci.sh file which changes the command of the yarn jest that uses the additional --coverage parameter, so that we can generate a coverage report for that specific package since in the ci.sh we have there has a command that will run jest

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (aldousalvarez) [reply .github/workflows/ci.yaml:2720]: @petermetz I chose cp-bungee-hermes because when we try to generate packages it is running simultaneously on all the step when they run tools.ci.sh it has the longest time that is passing when running tools.ci.sh because we need to make sure all the packages should generate a coverage report first before we merged them all because when I tried to depend on check-coverage only some of the packages does not get included only the finished ones are being covered by the code coverage report.

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (aldousalvarez) [reply .github/workflows/ci.yaml:2756]: @petermetz okay got it, I will work on this one and let you know as soon as it is done. Thank you

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (jenniferlianne) [reply .github/workflows/ci.yaml:2720]: If the final job was in a separate workflow triggered from ci.yaml, then you wouldn't need to worry about a longer test job being introduced.  

https://stackoverflow.com/questions/58457140/dependencies-between-workflows-on-github-actions

While I'm new to github, I get the impression artifacts are shared with a workflow triggered in sequence if they are uploaded by the upstream workflow.

https://stackoverflow.com/questions/60355925/share-artifacts-between-workflows-github-actions

Comment (aldousalvarez) [reply .github/workflows/ci.yaml:2720]: @jenniferlianne oh that it nice. I will take a look on this one and try it out. Thanks

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (petermetz) [reply .github/workflows/ci.yaml]: @aldousalvarez Thank you, this part of it looks good now.

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (petermetz) [reply .github/workflows/ci.yaml]: @aldousalvarez 

> @petermetz I chose cp-bungee-hermes because when we try to generate packages it is running simultaneously on all the step when they run tools.ci.sh it has the longest time that is passing when running tools.ci.sh because we need to make sure all the packages should generate a coverage report first before we merged them all because when I tried to depend on check-coverage only some of the packages does not get included only the finished ones are being covered by the code coverage report.

@aldousalvarez Unfortunately, that is not an optimal solution because the time each job takes in relation to each other is too varied on account of GitHub being a multi-tenant application with shared CPUs between jobs of ours and other people working on other projects. There are also a lot of other factors in play that changes (basically randomizes) the time jobs take to execute such as network congestion while fetching dependencies. For the CPU usage part, you can read more about the theory behind it if you look up the noisy neighbors[1] problem for example.

Long story short: I agree with @jenniferlianne 's excellent suggestion, please refactor it so that the ci.yaml workflow's conclusion triggers the code coverage which can be in a separate workflow file that could be named just `CI` and the job within it could be called `coverage_typescript`. We need to specify that it's typescript coverage because in the future we plan on having coverage for all the different programming languages that are present in the mono-repo (Kotlin, Rust, Go, Solidity etc.)

[1] https://learn.microsoft.com/en-us/azure/architecture/antipatterns/noisy-neighbor/noisy-neighbor

Comment (aldousalvarez) [reply .github/workflows/ci.yaml]: @petermetz Okay, Got it. Will do it and I am currently in progress of refactoring the code and testing how to integrate it. Thank you

**Comment by aldousalvarez** [comment]
Hello @petermetz already done with the requested changes. Thank you. It will not run on this pull request but it will run on the default branch which is main once it gets merged.

**Review by petermetz — APPROVED** [review]
> Hello @petermetz already done with the requested changes. Thank you. It will not run on this pull request but it will run on the default branch which is main once it gets merged.

@aldousalvarez Thank you for the updates! I agree it's time for us to take a look at it and see if it works on `main`. Fingers crossed! :-)

**Comment by jagpreetsinghsasan** [comment]
LGTM too

**Review by outSH — APPROVED** [review]

**Review by petermetz — CHANGES_REQUESTED** [review]
@aldousalvarez Please fix the merge conflicts and then we are good to go! (Make sure to pass it back for review once the conflicts are fixed)

**Comment by aldousalvarez** [comment]
@petermetz Already fixed the merge conflict. Thank you!

**Review by petermetz — CHANGES_REQUESTED** [review]
@aldousalvarez It might've happened again with another PR in the meantime, please fix again :-)

image

**Comment by aldousalvarez** [comment]
Hello @petermetz just fixed the merged conflicts. Thank you

**Review by petermetz — APPROVED** [review]
@aldousalvarez LGTM, thank you!
